### PR TITLE
add backtesting subcommand and refresh test data

### PR DESCRIPTION
**Pull request by gcarq**
This PR allows to execute backtesting as a subcommand i.e.: `freqtrade backtesting`
One major change is that is uses the current config.json instead of the built-in fixture. The configuration file can be changed with `-c, --config`.

Possible arguments for backtesting are:
```
usage: freqtrade backtesting [-h] [-l] [-i INT]

optional arguments:
  -h, --help            show this help message and exit
  -l, --live            using live data
  -i INT, --ticker-interval INT
                        specify ticker interval in minutes (default: 5)
```

This PR also contains a rewritten backtesting report using tabulate and new backtesting data for `oneMin` and `fiveMin` intervals.

Example usage
```
$ /usr/bin/freqtrade backtesting --ticker-interval 5
============================= test session starts ==============================
platform linux -- Python 3.6.3, pytest-3.2.3, py-1.4.34, pluggy-0.4.0
rootdir: /home/user/workspace/freqtrade, inifile:
plugins: mock-1.6.3, cov-2.5.1
collected 1 item

freqtrade/tests/test_backtesting.py
Using config: config.json ...
Using ticker_interval: 5 ...
Using local backtesting data (ignoring whitelist in given config)...
Using stake_currency: BTC ...
Using stake_amount: 0.01 ...
Measuring data from 2017-10-25T23:05:00+00:00 up to 2017-11-14T23:00:00+00:00 ...
====================== BACKTESTING REPORT ======================================

NOTE: This Report doesn't respect the limits of max_open_trades,
      so the projected values should be taken with a grain of salt.

pair         buy count  avg profit    total profit       avg duration
---------  -----------  ------------  ---------------  --------------
BTC_BCC              2  1.83%         0.03650892 BTC           522.5
BTC_ETH             46  0.17%         0.07942604 BTC           602.61
BTC_MER              6  2.70%         0.16224435 BTC            40
BTC_POWR             0  nan%          0.00000000 BTC           nan
BTC_ETC             84  0.96%         0.80512751 BTC           601.9
BTC_OK             101  1.30%         1.31318832 BTC           371.73
BTC_NEO            141  0.92%         1.29999502 BTC           443.37
BTC_EMC2            94  1.93%         1.81445864 BTC          1200.21
BTC_DASH            47  0.81%         0.37953663 BTC           567.66
BTC_LSK            130  0.94%         1.22181946 BTC           895.08
BTC_LTC            117  0.68%         0.79580105 BTC          1685.3
BTC_XZC            121  1.61%         1.94694915 BTC           698.35
BTC_OMG            164  1.19%         1.95760021 BTC          1517.16
BTC_STRAT          110  1.07%         1.17599907 BTC            87.86
BTC_XRP            115  1.43%         1.64984763 BTC           677.22
BTC_QTUM           126  0.90%         1.13416580 BTC          1261.75
BTC_WAVES          130  0.13%         0.17012238 BTC           501.5
BTC_VTC             86  2.12%         1.82154036 BTC           112.56
BTC_XLM            106  0.47%         0.49609272 BTC           512.31
BTC_MCO             91  0.33%         0.30380778 BTC           188.85
TOTAL             1817  1.02%         18.56423103 BTC          747.86
.

========================== 1 passed in 55.99 seconds ===========================

Process finished with exit code 0
```

### CODE REVIEW

**Comment by vertti** [comment]
This looks really nice. I like the new report layout. Hopefully it's not too difficult to merge with #111 .

File: freqtrade/main.py [hunk 1]
```diff
@@ -314,6 +314,11 @@ def main():
     global _CONF
     args = build_arg_parser().parse_args()
 
+    # Check if subcommand has been selected
+    if hasattr(args, 'func'):
+        args.func(args)
+        exit(0)
+
```
Comment (vertti) [inline freqtrade/main.py]: maybe we should push all the args stuff to separate method for unit testing etc

File: freqtrade/misc.py [hunk 2]
```diff
@@ -88,9 +89,49 @@ def build_arg_parser() -> argparse.ArgumentParser:
...
+        help='using live data',
+    )
+    backtest.add_argument(
+        '-i', '--ticker-interval',
+        help='specify ticker interval in minutes (default: 5)',
+        dest='ticker_interval',
+        default=5,
+        type=int,
+        metavar='INT',
+    )
```
Comment (vertti) [inline freqtrade/misc.py]: this should probably have some simple unit test?

File: freqtrade/misc.py [hunk 3]
```diff
@@ -88,9 +89,49 @@ def build_arg_parser() -> argparse.ArgumentParser:
...
+    import pytest
+
+    os.environ.update({
+        'BACKTEST': 'true',
+        'BACKTEST_LIVE': 'true' if args.live else '',
+        'BACKTEST_CONFIG': args.config,
+        'BACKTEST_TICKER_INTERVAL': str(args.ticker_interval),
+    })
+    path = os.path.join(os.path.dirname(__file__), 'tests', 'test_backtesting.py')
+    pytest.main(['-s', path])
```
Comment (vertti) [inline freqtrade/misc.py]: this should also have a simple unit test that checks the generated path and updated environment.

File: freqtrade/tests/__init__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+def load_backtesting_data(ticker_interval: int = 5):
+    path = os.path.abspath(os.path.dirname(__file__))
+    result = {}
+    pairs = [
+        'BTC_BCC', 'BTC_ETH', 'BTC_MER', 'BTC_POWR', 'BTC_ETC',
+        'BTC_OK', 'BTC_NEO', 'BTC_EMC2', 'BTC_DASH', 'BTC_LSK',
+        'BTC_LTC', 'BTC_XZC', 'BTC_OMG', 'BTC_STRAT', 'BTC_XRP',
+        'BTC_QTUM', 'BTC_WAVES', 'BTC_VTC', 'BTC_XLM', 'BTC_MCO'
+    ]
```
Comment (vertti) [inline freqtrade/tests/__init__.py]: I hope 20 coins is not slowing hyperopting too badly.

File: freqtrade/tests/test_backtesting.py [hunk 5]
```diff
@@ -1,36 +1,84 @@
...
-logging.disable(logging.DEBUG)  # disable debug logs that slow backtesting a lot
 
+def format_results(results: DataFrame):
+    return 'Made {} buys. Average profit {:.2f}%. ' \
+           'Total profit was {:.3f}. Average duration {:.1f} mins.'.format(
+               len(results.index),
+               results.profit.mean() * 100.0,
+               results.profit.sum(),
+               results.duration.mean() * 5,
+           )
```
Comment (vertti) [inline freqtrade/tests/test_backtesting.py]: I would actually prefer to have simple unit tests for these too. It seems trivial but at the same time it's important that we don't make mistakes and show let's say the wrong total profit.

File: freqtrade/tests/test_backtesting.py [hunk 6]
```diff
@@ -54,12 +102,43 @@ def backtest(backtest_conf, backdata, mocker):
...
+    config = config or backtest_conf
+
+    print('Using stake_currency: {} ...\nUsing stake_amount: {} ...'.format(
+        config['stake_currency'], config['stake_amount']
+    ))
+
+    min_date, max_date = get_timeframe(data)
+    print('Measuring data from {} up to {} ...'.format(
+        min_date.isoformat(), max_date.isoformat()
+    ))
```
Comment (vertti) [inline freqtrade/tests/test_backtesting.py]: this function is growing too large.

**Comment by gcarq** [comment]
@vertti I've added some tests to cover the argparse handling and backtesting. I will rebase it on develop.

File: freqtrade/tests/__init__.py (hunk 4 above)
Comment (gcarq) [reply freqtrade/tests/__init__.py]: I reduced the number of coins to 10

File: freqtrade/tests/test_backtesting.py (hunk 6 above)
Comment (gcarq) [reply freqtrade/tests/test_backtesting.py]: I created a shared function to load the config file, but I don't think we can reduce it more than that. The complexity is very low and half of the code contains logging

File: freqtrade/tests/test_misc.py [hunk 7]
```diff
@@ -18,3 +22,99 @@ def func():
...
+        'BACKTEST': 'true',
+        'BACKTEST_LIVE': 'true',
+        'BACKTEST_CONFIG': 'config.json',
+        'BACKTEST_TICKER_INTERVAL': '1',
+    }
+    assert pytest_mock.call_count == 1
+
+    main_call_args = pytest_mock.call_args[0][0]
+    assert main_call_args[0] == '-s'
+    assert main_call_args[1].endswith('freqtrade/tests/test_backtesting.py')
```
Comment (vertti) [inline freqtrade/tests/test_misc.py]: This fails on Windows machines as the path has different separator.

```
>       assert main_call_args[1].endswith('freqtrade/tests/test_backtesting.py')
E       AssertionError: assert False
E        +  where False = <built-in method endswith of str object at 0x000002BBC9DB0288>('freqtrade/tests/test_backtesting.py')
E        +    where <built-in method endswith of str object at 0x000002BBC9DB0288> = 'C:\\Users\\vertti\\dev\\freqtrade\\freqtrade\\tests\\test_backtesting.py'.endswith

freqtrade\tests\test_misc.py:120: AssertionError
```

Comment (vertti) [reply freqtrade/tests/test_misc.py]: Nevermind, I fixed it in 315919c

**Pull request merged**